### PR TITLE
Connection

### PR DESCRIPTION
**Pull request by mohe2015**
To track progress etc

Fixes:

- https://github.com/reform-org/reform/issues/114
- https://github.com/reform-org/reform/issues/131
- https://github.com/reform-org/reform/issues/139
- https://github.com/reform-org/reform/issues/154
- https://github.com/reform-org/reform/issues/153
- https://github.com/reform-org/reform/issues/165

### CODE REVIEW

File: js/src/main/scala/webapp/services/DiscoveryService.scala [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    console.log(name, payload)
+    name match {
+      case "request_host_token" => {
+        var iceServers = js.Array[RTCIceServer]()
+        iceServers += RTCIceServer(
+          Globals.turnServerURL,
+          payload.host.turn.username.asInstanceOf[String],
+          payload.host.turn.credential.asInstanceOf[String],
+        )
+        // iceServers += RTCIceServer("stun:lukasschreiber.com:41720")
```
Comment (lukasschreiber) [inline js/src/main/scala/webapp/services/DiscoveryService.scala:181]: comment can be removed, stun is never needed, the turn url is enough and can provide stun services without relaying just as good

File: js/src/main/scala/webapp/services/DiscoveryService.scala [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+          })
+      }
+      case "request_client_token" => {
+        var iceServers = js.Array[RTCIceServer]()
+        iceServers += RTCIceServer(
+          Globals.turnServerURL,
+          payload.client.turn.username.asInstanceOf[String],
+          payload.client.turn.credential.asInstanceOf[String],
+        )
+        // iceServers += RTCIceServer("stun:lukasschreiber.com:41720")
```
Comment (lukasschreiber) [inline js/src/main/scala/webapp/services/DiscoveryService.scala:210]: can be removed

File: js/src/main/scala/webapp/services/DiscoveryService.scala [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+      ws.get.onclose = (event) => {
+        setOnlineStatus.fire(false)
+        console.log("closed websocket")
+      }
+
+      ws.get.onerror = (event) => {
+        promise.failure(new Exception("Connection failed"))
+      }
+    } else {
+      promise.failure(new Exception("Something went wrong"))
```
Comment (lukasschreiber) [inline js/src/main/scala/webapp/services/DiscoveryService.scala:301]: set a better error message, we know that either the token was wrong or autoconnect was disabled

File: package.json [hunk 4]
```diff
@@ -31,6 +31,7 @@
     "chance": "^1.1.9",
     "daisyui": "^2.49.0",
     "idb-keyval": "^6.2.0",
+    "jsonwebtoken": "^9.0.0",
```
Comment (lukasschreiber) [inline package.json:34]: I think I don't need jsonwebtoken, I am now doing the decoding manually in Scala as far as I know

File: shared/src/main/scala/webapp/webrtc/WebRTCService.scala [hunk 5]
```diff
@@ -17,7 +17,103 @@ package webapp.webrtc
...
+import loci.communicator.webrtc.WebRTC.ConnectorFactory
+import com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec
+import com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker
+import webapp.utils.Base64
+import com.github.plokhotnyuk.jsoniter_scala.core.*
+import scala.scalajs.js
+import webapp.npm.Utils
+
+class ConnectionInformation(val session: WebRTC.CompleteSession, val alias: String, val source: String = "manual") {}
+class StoredConnectionInformation(
```
Comment (lukasschreiber) [inline shared/src/main/scala/webapp/webrtc/WebRTCService.scala:38]: Different classes for automatic and manual connections

Comment (lukasschreiber) [reply shared/src/main/scala/webapp/webrtc/WebRTCService.scala:38]: name EstablishedConnection to match the wording in the discovery server

File: shared/src/main/scala/webapp/webrtc/WebRTCService.scala [hunk 6]
```diff
@@ -17,7 +17,103 @@ package webapp.webrtc
...
+  def sessionAsToken(s: ConnectionInformation) = Base64.encode(writeToString(s)(codec))
+
+  def tokenAsSession(s: String) = readFromString(Base64.decode(s))(codec)
+}
 
 class WebRTCService {
   val registry: Registry = new Registry
+
+  private val connectionInfo = scala.collection.mutable.Map[RemoteRef, StoredConnectionInformation]()
+  private val webRTCConnections =
```
Comment (lukasschreiber) [inline shared/src/main/scala/webapp/webrtc/WebRTCService.scala:66]: connection Object should be part of "EstablishedConnection" aka StoredConnectionInfo

File: shared/src/main/scala/webapp/webrtc/WebRTCService.scala [hunk 7]
```diff
@@ -17,7 +17,103 @@ package webapp.webrtc
...
+  }
+
+  def closeConnectionById(id: String) = {
+    connectionRefs.get(id) match {
+      case None      => {}
+      case Some(ref) => ref.disconnect()
+    }
+  }
+
+  def getInformation(ref: RemoteRef): StoredConnectionInformation = {
```
Comment (lukasschreiber) [inline shared/src/main/scala/webapp/webrtc/WebRTCService.scala:106]: rename to be more clear, in the beginning it should only return Information aka the alias, now it should returnthe entire connection object as well es some additional information so getConnection should be more fitting

File: utils.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+export const usesTurn = async (connection) => {
+    const stats = await connection.getStats();
+    let selectedLocalCandidate;
+    for (const { type, state, localCandidateId } of stats.values())
+        if (type === 'candidate-pair' && state === 'succeeded' && localCandidateId) {
+            selectedLocalCandidate = localCandidateId;
+            break;
+        }
+    const usesTurn = !!selectedLocalCandidate && stats.get(selectedLocalCandidate)?.candidateType === 'relay';
+    return usesTurn;
```
Comment (lukasschreiber) [inline utils.js:10]: inline usesTurn

**Review by lukasschreiber — CHANGES_REQUESTED** [review]

**Comment by lukasschreiber** [comment]
- [x] Show all available remote connections
- [x] Allow user to login
- [x] handle no Internet
- [x] Allow user to whitelist clients in the UI (and redact that choice)
- [x] Make automatic connections able to be turned off
- [x] Extract URLs into scala file with globals
- [x] notify clients if connection has been lost --> close connection if it still exists
- [x] Show the user if connection is relay or local
- [x] Fix Dropdown - maybe make it a drawer
- [x] make Connection and Source a different color in connection dropdown
- [ ] Manual connect aliases are swapped
- [x] Error messages on login error
- [ ] show trusted and online clients that have been disconnected
- [ ] remove connection on revoke trust
- [x] Remove the untrust button from your own account
- [ ] Disconnecting peers that are already disconnected leads to an error
- [ ] User should see his username that is currently logged in

File: js/src/main/scala/webapp/Routes.scala [hunk 9]
```diff
@@ -42,11 +41,10 @@ object Routes {
     case ProjectsPage()          => Root / "projects"
     case UsersPage()             => Root / "users"
     case HiwisPage()             => Root / "hiwis"
-    case WebRTCHandling(_)       => Root / "webrtc";
     case PaymentLevelsPage()     => Root / "paymentlevels";
     case SalaryChangesPage()     => Root / "salarychanges"
     case SupervisorsPage()       => Root / "supervisor"
```
[the remark this answers is not in the document]
Comment (mohe2015) [reply js/src/main/scala/webapp/Routes.scala:46]: The autoformatter is not able to enforce this. I still removed them.

**Review by lukasschreiber — APPROVED** [review]

**Pull request merged**